File: packages/registro/src/usecases/CriarNovoRegistro.ts
import { CasoDeUso } from "common";
import { RegistroDTO } from "../model";
import { BackendRegistro } from "../provider";

// type Entrada = { registro: Partial<Registro> };
// type Entrada = { registro: Omit<Registro, "id"> };
type Entrada = { registro: RegistroDTO };

export class CriarNovoRegistro implements CasoDeUso<Entrada, Response | void> {
  public constructor(private backend: BackendRegistro) {}
  async executar(dto: Entrada): Promise<Response | void> {
    return await this.backend.criar(dto.registro);
  }
}

File: apps/backend-springboot/src/main/java/br/com/dinheironapalmadamao/backend_springboot/core/registro/usecases/CriarNovoRegistro.java
package br.com.dinheironapalmadamao.backend_springboot.core.registro.usecases;

import java.util.concurrent.CompletableFuture;

import br.com.dinheironapalmadamao.backend_springboot.core.common.CasoDeUso;
import br.com.dinheironapalmadamao.backend_springboot.core.registro.model.RegistroDTO;
import br.com.dinheironapalmadamao.backend_springboot.core.registro.provider.BackendRegistro;

public class CriarNovoRegistro implements CasoDeUso<RegistroDTO, Void> {
  private BackendRegistro backend;

  public CriarNovoRegistro(BackendRegistro backend) {
    this.backend = backend;
  }

  @Override
  public CompletableFuture<Void> excutar(RegistroDTO dto) {
    return CompletableFuture.runAsync(() -> {
      this.backend.criar(dto);
    });
  }

}
